File: ts-pr/55/typescript/code-quality-logging/request_handler.ts
class RequestHandler {
    handleRequest(method: string, url: string, params: Record<string, string>): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Request: ${method} ${url}`);
        Object.entries(params).forEach(([key, value]) => {
            console.log(`Param: ${key}=${value}`);
        });
        // {/fact}
    }

    secureHandleRequest(method: string, url: string, params: Record<string, string>): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Request: ${method} ${url}`);
        console.log(`Parameters count: ${Object.keys(params).length}`);
        // {/fact}
    }

    logRequestBody(body: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Request body: ${body}`);
        // {/fact}
    }

    secureLogRequestBody(bodySize: number): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Request body size: ${bodySize} bytes`);
        // {/fact}
    }

    logClientInfo(ip: string, userAgent: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Client: IP=${ip}, UserAgent=${userAgent}`);
        // {/fact}
    }

    secureLogClientInfo(hashedIP: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Client request from: ${hashedIP}`);
        // {/fact}
    }

    logCookies(cookies: Record<string, string>): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        Object.entries(cookies).forEach(([name, value]) => {
            console.log(`Cookie: ${name}=${value}`);
        });
        // {/fact}
    }

    secureLogCookies(cookieCount: number): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`Request contains ${cookieCount} cookies`);
        // {/fact}
    }
}

const handler = new RequestHandler();
const params = {
    'username': 'john_doe',
    'password': 'secret123'
};
handler.handleRequest('POST', '/login', params);

File: go-pr/48/go/code-quality-logging/request_handler.go
package main

import (
	"fmt"
	"log"
)

type RequestHandler struct{}

func (rh *RequestHandler) HandleRequest(method, url string, params map[string]string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("Request: %s %s\n", method, url)
	for key, value := range params {
		fmt.Printf("Param: %s=%s\n", key, value)
	}
	// {/fact}
}

func (rh *RequestHandler) SecureHandleRequest(method, url string, params map[string]string) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("Request: %s %s\n", method, url)
	log.Printf("Parameters count: %d\n", len(params))
	// {/fact}
}

func (rh *RequestHandler) LogRequestBody(body string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("Request body: %s\n", body)
	// {/fact}
}

func (rh *RequestHandler) SecureLogRequestBody(bodySize int) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("Request body size: %d bytes\n", bodySize)
	// {/fact}
}

func (rh *RequestHandler) LogClientInfo(ip, userAgent string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("Client: IP=%s, UserAgent=%s\n", ip, userAgent)
	// {/fact}
}

func (rh *RequestHandler) SecureLogClientInfo(hashedIP string) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("Client request from: %s\n", hashedIP)
	// {/fact}
}

func main() {
	handler := &RequestHandler{}
	params := map[string]string{
		"username": "john_doe",
		"password": "secret123",
	}
	handler.HandleRequest("POST", "/login", params)
}